packages/data-service/src/data-service.ts

```diff
@@ -2285,6 +2314,48 @@ class DataServiceImpl extends WithLogContext implements DataService {
     return { name, ...normalized };
   }
 
+  @op(mongoLogId(1_001_000_063))
+  async previewUpdate(
+    ns: string,
+    filter: Document,
+    update: Document
+  ): Promise<UpdatePreview> {
+    const coll = this._collection(ns, 'CRUD');
+    const session = this._startSession('CRUD');
+    try {
+      session.startTransaction();
+
+      const docsToPreview = await coll
+        .find(filter, { session })
+        .sort({ _id: 1 })
+        .limit(PREVIEW_SAMPLE)
+        .maxTimeMS(PREVIEW_TIMEOUT)
+        .toArray();
+
+      const idsToPreview = docsToPreview.map((doc) => doc._id);
+      await coll.updateMany({ _id: { $in: idsToPreview } }, update, {
```

Comment:
Maybe I’m just unnecessarily scared by our dots-and-dollars support, but maybe let’s test this with documents that look like
```js
{
  _id: {
    'foo.bar': 42
  }
}
```

(This is a valid document, and this code seems to support it just fine, it just feels like the kind of thing where at some point somebody will do something that will optimize the `$in` away, replace it by something else, and then break this edge case…)